### ISSUE TITLE
CUDA must be available in order to use Big Sleep

### ISSUE THREAD

**Issue by calebrader** [lucidrains/big-sleep#80]
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Sun_Feb_14_22:08:44_Pacific_Standard_Time_2021
Cuda compilation tools, release 11.2, V11.2.152
Build cuda_11.2.r11.2/compiler.29618528_0

**Comment by wolfgangmeyers** [comment]
I ran into this on windows. I had to custom install pytorch from this page: https://pytorch.org/get-started/locally/

**Comment by LoudMurmur** [comment]
yeah same, you have to install pytorch before doing pip install big_sleep

**Comment by jimleonardo** [comment]
> I ran into this on windows. I had to custom install pytorch from this page: https://pytorch.org/get-started/locally/

This is what I needed to do with a GeForce RTX 2070. I installed CUDA 10.2 for now because it seemed surest bet just to get up and running. Now big-sleep is generating for me.

**Comment by HenryFBP** [comment]
> > I ran into this on windows. I had to custom install pytorch from this page: https://pytorch.org/get-started/locally/
> 
> This is what I needed to do with a GeForce RTX 2070. I installed CUDA 10.2 for now because it seemed surest bet just to get up and running. Now big-sleep is generating for me.

Windows 11 Pro, Lenovo Legion 82JY Laptop, RTX 3060 GPU here.

I did this, by running `conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge`. I had to:

- uninstall python
- install conda
- run (admin) `conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge`
- run `pip install big-sleep`
- run `dream happy sheep`

Although now I get a new problem:

```

C:\Users\henryfbp> dream happy sheep
Key already registered with the same priority: GroupSpatialSoftmax```

**Comment by halr9000** [comment]
> Make sure you're on Python 3.8 btw, 3.9 & 3.10 don't work.

Not sure if it was me reinstalling cuda or bumping down to Python 3.8 which finally worked. Do we know why, or do you have a reference for that @ThatOneCalculator ?

**Comment by ThatOneCalculator** [comment]
> > Make sure you're on Python 3.8 btw, 3.9 & 3.10 don't work.
> 
> Not sure if it was me reinstalling cuda or bumping down to Python 3.8 which finally worked. Do we know why, or do you have a reference for that @ThatOneCalculator ?

It's because Python 3.9 broke the cuda package.

**Comment by illtellyoulater** [comment]
What worked for me on Windows was:

- deactivating and removing previously created Big Sleep conda environment with `conda deactivate bigsleep` and `conda remove -n bigsleep --all`
- deleting bigsleep env folder at `c:\users\my-username\.conda\envs\bigsleep`
- creating and activating a new conda environment featuring python 3.8, with `conda create --name bigsleep python=3.8` and `conda activate bigsleep`
- installing pytorch & cudatoolkit with `conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch` (as per current installation instructions at https://pytorch.org/get-started/locally/ for Windows systems)
- installing Big Sleep with `pip install big-sleep`

Hope it helps

**Comment by calebrader** [comment]
How do I install Conda for Python 3.8 ?

**Comment by illtellyoulater** [comment]
> How do I install Conda for Python 3.8 ?

Just download the full Conda distribution from here https://www.anaconda.com/products/individual

There could be ways to add Conda to an already working Python env but the above approach will work more consistently I believe.

**Comment by calebrader** [comment]
I followed the instructions from above and got this error:

```
(base) C:\Users\caleb>dream happy sheep
Traceback (most recent call last):
  File "C:\Users\caleb\anaconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\caleb\anaconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\caleb\anaconda3\Scripts\dream.exe\__main__.py", line 4, in <module>
  File "C:\Users\caleb\anaconda3\lib\site-packages\big_sleep\__init__.py", line 1, in <module>
    from big_sleep.big_sleep import BigSleep, Imagine
  File "C:\Users\caleb\anaconda3\lib\site-packages\big_sleep\big_sleep.py", line 26, in <module>
    assert torch.cuda.is_available(), 'CUDA must be available in order to use Big Sleep'
AssertionError: CUDA must be available in order to use Big Sleep
```

**Comment by calebrader** [comment]
How do I get rid of python 3.9? 

```
(bigsleep) C:\Users\caleb>py --version
Python 3.9.7

(bigsleep) C:\Users\caleb>
```

**Comment by meltingscales** [comment]
> How do I get rid of python 3.9?
> 
> ```
> (bigsleep) C:\Users\caleb>py --version
> Python 3.9.7
> 
> (bigsleep) C:\Users\caleb>
> ```

I think you might be in a virtual env, you can run `where.exe py` to see where the `py` command/exe comes from. I use Pipenv so I only know `pipenv --py` to show environment info. Perhaps conda(?) has a similar command.

**Comment by calebrader** [comment]
> > How do I install Conda for Python 3.8 ?
> 
> Just download the full Conda distribution from here https://www.anaconda.com/products/individual
> 


This doesn't work. The "full Conda distribution" currently uses Python 3.9 which breaks Cuda and prevents Big Sleep from running.

**Comment by meltingscales** [comment]
> > > How do I install Conda for Python 3.8 ?
> > 
> > 
> > Just download the full Conda distribution from here https://www.anaconda.com/products/individual
> 
> This doesn't work. The "full Conda distribution" currently uses Python 3.9 which breaks Cuda and prevents Big Sleep from running.

Maybe you could try an older Conda release? I will check what version I have on the windows laptop running CUDA.

**Comment by halr9000** [comment]
Quoting @calebrader:
> This doesn't work. The "full Conda distribution" currently uses Python 3.9 which breaks Cuda and prevents Big Sleep from running.

You are in luck, conda handles package management even of Python itself, and @illtellyoulater actually provides an answer [above ](https://github.com/lucidrains/big-sleep/issues/80#issuecomment-1061281157)that shows this in action.

Look at the [conda docs](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-python.html#:~:text=Conda%20treats%20Python%20the%20same,%2C%22%20the%20default%20is%202.7.) where they explain in more detail.

tl;dr re-read his message and try that. The line that does the trick is this one:

`conda create --name bigsleep python=3.8`
